esiclient/v1/orchestrator/openshift.py

```diff
@@ -0,0 +1,339 @@
+    """Orchestrate an OpenShift cluster using the Assisted Installer API"""
+
+    log = logging.getLogger(__name__ + ".Orchestrate")
+
+    def _print_failure_message(self, cluster_config_file, cluster_id=None,
+                               infra_env_id=None, message=None):
+        flags = ""
+        if cluster_id:
+            flags = "--cluster-id %s" % cluster_id
+            if infra_env_id:
+                flags = "%s --infra-env-id %s" % (flags, infra_env_id)
+        command = "openstack esi orchestrate openshift %s %s" % \
+            (flags, cluster_config_file)
+        if message:
+            print(message)
+        print("You may need to refresh your OpenShift token.")
+        print("Run this command to continue installation: %s" % command)
+        return
+
+    def get_parser(self, prog_name):
+        parser = super(Orchestrate, self).get_parser(prog_name)
+
+        parser.add_argument(
+            "cluster_config_file",
+            metavar="<cluster_config_file>",
+            help=_("File describing the cluster configuration"))
+        parser.add_argument(
+            "--cluster-id",
+            metavar="<cluster_id>",
+            help=_("OpenShift cluster ID"))
+        parser.add_argument(
+            "--infra-env-id",
+            metavar="<infra_env_id>",
+            help=_("OpenShift infrastruction environment ID"))
+
+        return parser
+
+    def take_action(self, parsed_args):
+        self.log.debug("take_action(%s)", parsed_args)
+
+        cluster_id = parsed_args.cluster_id
+        infra_env_id = parsed_args.infra_env_id
+
+        with open(parsed_args.cluster_config_file) as f:
+            cluster_config = json.load(f)
+        nodes = cluster_config.get('nodes')
+        cluster_name = cluster_config.get('cluster_name')
+        provisioning_network_name = cluster_config.get(
+            'provisioning_network_name', 'provisioning')
+        external_network_name = cluster_config.get('external_network_name')
+        private_network_name = cluster_config.get('private_network_name')
+        private_subnet_name = cluster_config.get('private_subnet_name')
+        api_vip = cluster_config.get('api_vip')
+        ingress_vip = cluster_config.get('ingress_vip')
+        openshift_version = cluster_config.get('openshift_version')
+        high_availability_mode = cluster_config.get(
+            'high_availability_mode', 'Full')
+        base_dns_domain = cluster_config.get('base_dns_domain')
+        ssh_public_key = cluster_config.get('ssh_public_key')
+
+        pull_secret = json.loads(os.environ['PULL_SECRET'])
+
+        ironic_client = self.app.client_manager.baremetal
+        neutron_client = self.app.client_manager.network
+
+        headers = {
+            'Content-Type': 'application/json',
+            'Authorization': 'Bearer ' + os.getenv('API_TOKEN', ''),
+        }
+
+        print("STARTING OPENSHIFT CLUSTER INSTALL")
+
+        # get cluster id
```

Comment:
Can you break down this take_action() method into smaller methods? Such as _get_cluster_id(), _register nodes(), _install_cluster(), etc.

Reply:
I actually tried that earlier, and it made the code really ugly, since a function such as _register_nodes will require a ton of openshift and openstack-relevant parameters.